### PR TITLE
maint: Move Go packages into root of repo, adopt go.work

### PR DESCRIPTION
**Pull request by mslipper**

### CODE REVIEW

File: go.work [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+	./l2geth
+	./l2geth-exporter
+	./op-exporter
+	./proxyd
+	./teleportr
+)
+
+replace github.com/ethereum-optimism/optimism/bss-core v0.0.0 => ./bss-core
+
+replace github.com/ethereum-optimism/optimism/l2geth v0.0.0 => ./l2geth
```
[the remark this answers is not in the document]
Comment (mslipper) [reply go.work:17]: You are correct. I'll drop this.

[the remark this answers is not in the document]
Comment (mslipper) [reply go.work:17]: Hmm, Go workspaces doesn't let me do that:

```
go: workspace module github.com/ethereum-optimism/optimism/l2geth is replaced at all versions in the go.work file
```

**Comment by mslipper** [comment]
Added do not merge so more people than 2 can approve.

File: README.md [hunk 2]
```diff
@@ -38,10 +38,10 @@ root
 │   ├── <a href="./packages/message-relayer">message-relayer</a>: Tool for automatically relaying L1<>L2 messages in development
 │   └── <a href="./packages/replica-healthcheck">replica-healthcheck</a>: Service for monitoring the health of a replica node
 ├── <a href="./go">go</a>
```
Comment (smartcontracts) [inline README.md:40]: If we want to be 100% correct, we should reformat this tree to match the new folder structure

File: batch-submitter/Dockerfile [hunk 3]
```diff
@@ -2,12 +2,12 @@ FROM golang:1.18.0-alpine3.15 as builder
 
 RUN apk add --no-cache make gcc musl-dev linux-headers git jq bash
 
-COPY ./l2geth /l2geth
-COPY ./go/bss-core /go/bss-core
-COPY ./go/batch-submitter/go.mod ./go/batch-submitter/go.sum /go/batch-submitter/
+COPY ./batch-submitter /go/batch-submitter
```
Comment (smartcontracts) [inline batch-submitter/Dockerfile:5]: Why do we copy into a `go` directory?

**Review by smartcontracts — APPROVED** [review]

File: batch-submitter/Dockerfile [hunk 4]
```diff
@@ -2,12 +2,12 @@ FROM golang:1.18.0-alpine3.15 as builder
...
+COPY ./bss-core /go/bss-core
+COPY ./l2geth /go/l2geth
+COPY ./batch-submitter/docker.go.work /go/go.work
+
 WORKDIR /go/batch-submitter
-RUN go mod graph | grep -v l2geth | grep -v bss-core | awk '{if ($1 !~ "@") print $2}' | xargs -n 1 go get
```
Comment (Inphi) [inline batch-submitter/Dockerfile:9]: That's cool!

**Review by Inphi — APPROVED** [review]
LGTM!
If it builds, it builds

File: README.md (hunk 2 above)
Comment (protolambda) [reply README.md:40]: This is being reformatted in the larger monorepo PR :+1:

File: batch-submitter/Dockerfile (hunk 3 above)
Comment (mslipper) [reply batch-submitter/Dockerfile:5]: It's just a nice container for all the go code - can be whatever we want.

**Pull request merged**